tss-esapi/examples/hmac.rs

```diff
@@ -0,0 +1,151 @@
+        SymmetricCipherParameters,
+        SymmetricDefinitionObject,
+    }
+};
+
+use std::convert::TryFrom;
+
+fn main() {
+    // Create a new TPM context. This reads from the environment variable `TPM2TOOLS_TCTI`.
+    //
+    // It's recommended you use `TPM2TOOLS_TCTI=device:/dev/tpmrm0` for the linux kernel
+    // tpm resource manager.
+    let mut context = Context::new(
+        TctiNameConf::from_environment_variable().expect("Failed to get TPM2TOOLS_TCTI"),
+    ).expect("Failed to create Context");
+
+    // Create the primary key. A primary key is the "root" of a collection of objects.
+    // These other objects are encrypted by the primary key allowing them to persist
+    // over a reboot and reloads.
+    //
+    // A primary key is derived from a seed, and provided that the same inputs are given
+    // the same primary key will be derived in the tpm. This means that you do not need
+    // to store or save the details of this key - only the parameters of how it was created.
+
+    let object_attributes = ObjectAttributesBuilder::new()
+        // Indicate the key can only exist within this tpm and can not be exported.
+        .with_fixed_tpm(true)
+        // The primary key and it's descendent keys can't be moved to other primary
+        // keys.
+        .with_fixed_parent(true)
+        // The primary key will persist over suspend and resume of the system.
+        .with_st_clear(false)
+        // The primary key was generated entirely inside the TPM - only this TPM
+        // knows it's content.
+        .with_sensitive_data_origin(true)
+        // This key requires "authentication" to the TPM to access - this can be
+        // an HMAC or password session. HMAC sessions are used by default with
+        // the "execute_with_nullauth_session" function.
+        .with_user_with_auth(true)
+        // This key has the ability to decrypt
+        .with_decrypt(true)
+        // This key may only be used to encrypt or sign objects that are within
+        // the TPM - it can not encrypt or sign external data.
+        .with_restricted(true)
+        .build()
+        .expect("Failed to build object attributes");
+
+    let primary_pub = PublicBuilder::new()
+        // This key is a symmetric key.
+        .with_public_algorithm(PublicAlgorithm::SymCipher)
+        .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+        .with_object_attributes(object_attributes)
+        .with_symmetric_cipher_parameters(
+            SymmetricCipherParameters::new(SymmetricDefinitionObject::AES_128_CFB)
+        )
+        .with_symmetric_cipher_unique_identifier(Digest::default())
+        .build()
+        .unwrap();
+
+    let primary = context.execute_with_nullauth_session(|ctx| {
+        // Create the key under the "owner" hierachy. Other hierachies are platform
+        // which is for boot services, null which is ephemeral and resets after a reboot,
+        // and endorsement which allows key certification.
```

Comment:
Small suggestion to make it a bit clearer:
```suggestion
        // and endorsement which allows key certification by the TPM manufacturer.
```